filters/fadein/fadein.go

```diff
@@ -0,0 +1,260 @@
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/routing"
+)
+
+const (
+	FadeInName          = "fadeIn"
+	EndpointCreatedName = "endpointCreated"
+)
+
+type (
+	fadeIn struct {
+		duration time.Duration
+		exponent float64
+	}
+
+	endpointCreated struct {
+		when  time.Time
+		which string
+	}
+
+	detectedFadeIn struct {
+		when       time.Time
+		duration   time.Duration
+		lastActive time.Time
+	}
+
+	postProcessor struct {
+		// "http://10.2.1.53:1234": {t0 60s t0-10s}
+		detected map[string]detectedFadeIn
+	}
+)
+
+// NewFadeIn creates a filter spec for the fade-in filter.
+func NewFadeIn() filters.Spec {
+	return fadeIn{}
+}
+
+func (fadeIn) Name() string { return FadeInName }
+
+func (fadeIn) CreateFilter(args []interface{}) (filters.Filter, error) {
+	if len(args) == 0 || len(args) > 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	var f fadeIn
+	switch v := args[0].(type) {
+	case int:
+		f.duration = time.Duration(v * int(time.Millisecond))
+	case float64:
+		f.duration = time.Duration(int(v) * int(time.Millisecond))
+	case string:
+		d, err := time.ParseDuration(v)
+		if err != nil {
+			return nil, err
+		}
+
+		f.duration = d
+	case time.Duration:
+		f.duration = v
+	default:
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	f.exponent = 1
+	if len(args) == 2 {
+		switch v := args[1].(type) {
+		case int:
+			f.exponent = float64(v)
+		case float64:
+			f.exponent = v
+		default:
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return f, nil
+}
+
+func (fadeIn) Request(filters.FilterContext)  {}
+func (fadeIn) Response(filters.FilterContext) {}
+
+// NewEndpointCreated creates a filter spec for the endpointCreated filter.
+func NewEndpointCreated() filters.Spec {
+	var ec endpointCreated
+	return ec
+}
+
+func (endpointCreated) Name() string { return EndpointCreatedName }
+
+func normalizeSchemeHost(s, h string) (string, string, error) {
+	// endpoint address cannot contain path, the rest is not case sensitive
+	s, h = strings.ToLower(s), strings.ToLower(h)
+
+	hh, p, err := net.SplitHostPort(h)
+	if err != nil {
+		// what is the actual right way of doing this, considering IPv6 addresses, too?
+		if !strings.Contains(err.Error(), "missing port") {
+			return "", "", err
+		}
+
+		p = ""
+	} else {
+		h = hh
+	}
+
+	switch {
+	case p == "" && s == "http":
+		p = "80"
+	case p == "" && s == "https":
+		p = "443"
+	}
+
+	h = net.JoinHostPort(h, p)
+	return s, h, nil
+}
+
+func normalizeEndpoint(e string) (string, string, error) {
+	u, err := url.Parse(e)
+	if err != nil {
+		return "", "", err
+	}
+
+	return normalizeSchemeHost(u.Scheme, u.Host)
+}
+
+func endpointKey(scheme, host string) string {
+	return fmt.Sprintf("%s://%s", scheme, host)
+}
+
+// endpointCreated("", <int,float64,string,time>)
+// endpointCreated("http://10.2.4.123:2156", $time)
```

Comment:
Better to have godoc format :)

Reply:
not even sure how this made it here, it might have been just some notes :)